src/rpc/tungstenite_client/client.rs

```diff
@@ -0,0 +1,330 @@
+/*
+   Copyright 2019 Supercomputing Systems AG
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+	   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+
+*/
+use crate::{
+	json_req,
+	rpc::{Error as RpcClientError, Result, RpcClient as RpcClientTrait},
+	tungstenite_client::{
+		read_until_text_message, GetRequestHandler, SubmitAndWatchHandler, SubmitOnlyHandler,
+		SubscriptionHandler,
+	},
+	FromHexString, HandleMessage, Subscriber, XtStatus,
+};
+use log::{debug, error, info, warn};
+use serde_json::Value;
+use sp_core::H256 as Hash;
+use std::{
+	fmt::Debug, net::TcpStream, sync::mpsc::Sender as ThreadOut, thread, thread::sleep,
+	time::Duration,
+};
+use tungstenite::{
+	client::connect_with_config,
+	protocol::{frame::coding::CloseCode, CloseFrame},
+	stream::MaybeTlsStream,
+	Message, WebSocket,
+};
+use url::Url;
+
+pub(crate) type MySocket = WebSocket<MaybeTlsStream<TcpStream>>;
+
+#[derive(Debug, Clone)]
+pub struct TungsteniteRpcClient {
+	url: String,
+	max_attempts: u8,
+}
+
+impl TungsteniteRpcClient {
+	pub fn new(url: &str, max_attempts: u8) -> TungsteniteRpcClient {
+		TungsteniteRpcClient { url: url.to_string(), max_attempts }
+	}
+
+	fn direct_rpc_request<MessageHandler>(
+		&self,
+		json_req: String,
+		message_handler: MessageHandler,
+	) -> Result<String>
+	where
+		MessageHandler: HandleMessage<Error = (RpcClientError, bool), Context = MySocket, Result = String>
+			+ Clone
+			+ Send
+			+ 'static,
+		MessageHandler::ThreadMessage: Send + Sync + Debug,
+	{
+		let url = Url::parse(self.url.as_str()).map_err(|e| RpcClientError::Other(e.into()))?;
+		let mut current_attempt: u8 = 1;
+		let mut socket: MySocket;
+
+		while current_attempt <= self.max_attempts {
+			match connect_with_config(url.clone(), None, 2) {
+				Ok(res) => {
+					socket = res.0;
+					let response = res.1;
+					debug!("Connected to the server. Response HTTP code: {}", response.status());
+					if socket.can_read() {
```

Comment:
This code is getting hard to read because a lot of ifs and loops are used. To avoid stacking `if { .. if { .. ` you could do the opposite, look for an error and return the error if found.
E.g. simply do 
```rust
socket.can_read()?;
// Check for ping message
socket.read_message()?; 
```
If you want to handle the error in a specific way, you could also write these checks in a separate function and handle the error here. Like
```rust
if let Err(e) = check_connection(&mut socket) {
    error!("adbds");
}
// continue..
```